Articles/Blog/2023-05-sci-workflows.md

```diff
@@ -0,0 +1,93 @@
+
+### Case study: Simulating tsunamis
+
+We wanted to perform Bayesian inference with uncertainty quantification (UQ) on a tsunami model, determining the tsunami source from buoy data. [ExaHyPE](https://doi.org/10.1016/j.cpc.2020.107251) served as the numerical model, while [MUQ](https://mituq.bitbucket.io/source/_site/index.html) provided the UQ method.
+
+Linking the two is - on a theoretical level - straightforward: MUQ's multilevel Markov chain Monte Carlo method only needs to iteratively pass parameter vectors to the tsunami model and receive some quantity of interest in return. Existing MPI support on both sides should allow for good scalability.
+
+<br> 
+
+<img src='../../images/sci-workflows-3.png' class='page lightbox' />[Monolithic application linking UQ algorithm and numerical model.]
+
+<br>
+
+In practice, it turned out to be a challenge: In order to optimize algorithms and software, we often have to intentionally restrict the set of supported use cases.
+
+In case of ExaHyPE, this meant focusing on a one-shot workflow where the user specifies the problem, and hardware-optimized, problem-specific code is generated and compiled. In addition, ExaHyPE builds on a fairly large number of dependencies and employs a somewhat unusual hybrid parallelization approach combining MPI with Intel's oneAPI [Threading Building Blocks](https://en.wikipedia.org/wiki/Threading_Building_Blocks) (TBB).
+
+Taken together, these properties meant that it took us months to implement a scalable application linking MUQ and ExaHyPE. We had to dig deep into the parallelization of ExaHyPE for it to accept MPI sub-communicators from MUQ, modify its main routines to call it as a library, and rewrite parts of its custom build system. Along the way, we encountered a myriad of unexpected issues.
+
+### Microservices in scientific software
+
+Linking MUQ and ExaHyPE was a painful experience, contrary to the theoretically simple link of the algorithms.
+
+We, therefore, decided to try a completely different approach, taking inspiration from microservice architectures. Enter UM-Bridge, the UQ and Modeling Bridge: Instead of building one monolithic application encompassing UQ, numerical model, and HPC capabilities, we now separate each aspect into its own component. The UQ application requests evaluations through the UM-Bridge network protocol, while the model application is now a server, launching simulation runs upon request. The model may run on a cluster, where a load balancer takes care of many parallelization aspects.
```

Comment:
@linusseelinger, I think it might be useful if you could add just a couple of sentences describing a little bit more about how UM-Bridge is "constructed" and how it works.  What language is it implemented in?  What is this "network protocol" you mention?  What is required to hook any given model/tool up with UM-Bridge?  Not huge detail, but I think it needs a bit more explanation.

Reply:
@bernhold I've added more detail on UM-bridge, do you think that gives sufficient insight?